File: tourcount/src/main/java/com/wmstein/tourcount/widgets/EditMetaWidget.java
package com.wmstein.tourcount.widgets;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.wmstein.tourcount.R;

import java.util.Objects;

/*************************************************************
 * EditMetaWidget.java used by EditMetaActivity.java
 * Created by wmstein for com.wmstein.tourcount on 2016-04-02,
 * last edited on 2019-02-12
 */
public class EditMetaWidget extends LinearLayout
{
    private final TextView widget_temp1; // temperature
    private final EditText widget_temp2;
    private final TextView widget_wind1; // wind
    private final EditText widget_wind2;
    private final TextView widget_clouds1; // clouds
    private final EditText widget_clouds2;
    private final TextView widget_plz1; // plz
    private final TextView widget_plz2;
    private final TextView widget_city1; // city
    private final TextView widget_city2;
    private final TextView widget_place1; // place
    private final TextView widget_place2;
    private final TextView widget_date1; // date
    private final TextView widget_date2;
    private final TextView widget_startTm1; // start_tm
    private final TextView widget_startTm2;
    private final TextView widget_endTm1; // end_tm
    private final TextView widget_endTm2;

    private final String regEx = "^[0-9]*$"; // plausi for numeric input

    public EditMetaWidget(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        Objects.requireNonNull(inflater).inflate(R.layout.widget_edit_meta, this, true);
        widget_temp1 = findViewById(R.id.widgetTemp1); // temperature
        widget_temp2 = findViewById(R.id.widgetTemp2);
        widget_wind1 = findViewById(R.id.widgetWind1); // wind
        widget_wind2 = findViewById(R.id.widgetWind2);
        widget_clouds1 = findViewById(R.id.widgetClouds1); // clouds
        widget_clouds2 = findViewById(R.id.widgetClouds2);
        widget_plz1 = findViewById(R.id.widgetPlz1); // plz
        widget_plz2 = findViewById(R.id.widgetPlz2);
        widget_city1 = findViewById(R.id.widgetCity); // city
        widget_city2 = findViewById(R.id.widgetItem4);
        widget_place1 = findViewById(R.id.widgetPlace); // place
        widget_place2 = findViewById(R.id.widgetItem5);
        widget_date1 = findViewById(R.id.widgetDate1);
        widget_date2 = findViewById(R.id.widgetDate2);
        widget_startTm1 = findViewById(R.id.widgetStartTm1);
        widget_startTm2 = findViewById(R.id.widgetStartTm2);
        widget_endTm1 = findViewById(R.id.widgetEndTm1);
        widget_endTm2 = findViewById(R.id.widgetEndTm2);
    }

    /**
     * Checks if a CharSequence is empty ("") or null.
     * <p>
     * isEmpty(null)      = true
     * isEmpty("")        = true
     * isEmpty(" ")       = false
     * isEmpty("bob")     = false
     * isEmpty("  bob  ") = false
     *
     * @param cs the CharSequence to check, may be null
     * @return {@code true} if the CharSequence is empty or null
     */
    private static boolean isEmpty(final CharSequence cs)
    {
        return cs == null || cs.length() == 0;
    }

    // Following the SETS
    // temperature
    public void setWidgetTemp1(String title)
    {
        widget_temp1.setText(title);
    }

    // wind
    public void setWidgetWind1(String title)
    {
        widget_wind1.setText(title);
    }

    // clouds
    public void setWidgetClouds1(String title)
    {
        widget_clouds1.setText(title);
    }

    // PLZ
    public void setWidgetPlz1(String title)
    {
        widget_plz1.setText(title);
    }

    // city
    public void setWidgetCity1(String title)
    {
        widget_city1.setText(title);
    }

    // place
    public void setWidgetPlace1(String title)
    {
        widget_place1.setText(title);
    }

    // date
    public void setWidgetDate1(String title)
    {
        widget_date1.setText(title);
    }

    // start_tm
    public void setWidgetStartTm1(String title)
    {
        widget_startTm1.setText(title);
    }

    // end_tm
    public void setWidgetEndTm1(String title)
    {
        widget_endTm1.setText(title);
    }

    // following the GETS
    // get temperature with plausi
    public int getWidgetTemp2()
    {
        String text = widget_temp2.getText().toString();
        if (isEmpty(text))
            return 0;
        else if (!text.trim().matches(regEx))
            return 100;
        else
        {
            try
            {
                return Integer.parseInt(text.replaceAll("[\\D]",""));
            } catch (NumberFormatException nfe)
            {
                return 100;
            }
        }
    }

    public void setWidgetTemp2(int name)
    {
        widget_temp2.setText(String.valueOf(name));
    }

    // get wind with plausi
    public int getWidgetWind2()
    {
        String text = widget_wind2.getText().toString();
        if (isEmpty(text))
            return 0;
        else if (!text.trim().matches(regEx))
            return 100;
        else
        {
            try
            {
                return Integer.parseInt(text.replaceAll("[\\D]",""));
            } catch (NumberFormatException nfe)
            {
                return 100;
            }
        }
    }

    public void setWidgetWind2(int name)
    {
        widget_wind2.setText(String.valueOf(name));
    }

    // get clouds with plausi
    public int getWidgetClouds2()
    {
        String text = widget_clouds2.getText().toString();
        if (isEmpty(text))
            return 0;
        else if (!text.trim().matches(regEx))
            return 200;
        else
        {
            try
            {
                return Integer.parseInt(text.replaceAll("[\\D]",""));
            } catch (NumberFormatException nfe)
            {
                return 100;
            }
        }
    }

    public void setWidgetClouds2(int name)
    {
        widget_clouds2.setText(String.valueOf(name));
    }

    // get PLZ with plausi
    public String getWidgetPlz2()
    {
        return widget_plz2.getText().toString();
    }

    public void setWidgetPlz2(String name)
    {
        widget_plz2.setText(name);
    }

    // get city with plausi
    public String getWidgetCity2()
    {
        return widget_city2.getText().toString();
    }

    public void setWidgetCity2(String name)
    {
        widget_city2.setText(name);
    }

    // get place with plausi
    public String getWidgetPlace2()
    {
        return widget_place2.getText().toString();
    }

    public void setWidgetPlace2(String name)
    {
        widget_place2.setText(name);
    }

    public String getWidgetDate2()
    {
        return widget_date2.getText().toString();
    }

    public void setWidgetDate2(String name)
    {
        widget_date2.setText(name);
    }

    public String getWidgetStartTm2()
    {
        return widget_startTm2.getText().toString();
    }

    public void setWidgetStartTm2(String name)
    {
        widget_startTm2.setText(name);
    }

    public String getWidgetEndTm2()
    {
        return widget_endTm2.getText().toString();
    }

    public void setWidgetEndTm2(String name)
    {
        widget_endTm2.setText(name);
    }

}

File: tourcount/src/main/java/com/wmstein/tourcount/widgets/EditMetaWidget.kt
package com.wmstein.tourcount.widgets

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import com.wmstein.tourcount.R
import java.util.Objects

/*************************************************************
 * EditMetaWidget.kt used by EditMetaActivity.java
 * Created by wmstein for com.wmstein.tourcount on 2016-04-02,
 * last edited in Java on 2019-02-12,
 * converted to Kotlin on 2023-07-09,
 * last edited on 2025-11-11
 */
class EditMetaWidget(context: Context, attrs: AttributeSet?) : LinearLayout(context, attrs) {
    // date
    private val widgetdate1: TextView
    private val widgetdate2: TextView

    // start time
    private val widgetstartTm1: TextView
    private val widgetstartTm2: TextView

    // end time
    private val widgetendTm1: TextView
    private val widgetendTm2: TextView

    // temperature
    private val widgettemp1: TextView
    private val widgettemp2: EditText
    private val widgettemp3: EditText

    // wind
    private val widgetwind1: TextView
    private val widgetwind2: EditText
    private val widgetwind3: EditText

    // clouds
    private val widgetclouds1: TextView
    private val widgetclouds2: EditText
    private val widgetclouds3: EditText

    val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    init {
        Objects.requireNonNull(inflater)
            .inflate(R.layout.widget_edit_meta, this, true)
        widgetdate1 = findViewById(R.id.widgetDate1)
        widgetdate2 = findViewById(R.id.widgetDate2)
        widgetstartTm1 = findViewById(R.id.widgetStartTm1)
        widgetstartTm2 = findViewById(R.id.widgetStartTm2)
        widgetendTm1 = findViewById(R.id.widgetEndTm1)
        widgetendTm2 = findViewById(R.id.widgetEndTm2)

        widgettemp1 = findViewById(R.id.widgetTemp1) // temperature
        widgettemp2 = findViewById(R.id.widgetStartTemp)
        widgettemp3 = findViewById(R.id.widgetEndTemp)
        widgetwind1 = findViewById(R.id.widgetWind1) // wind
        widgetwind2 = findViewById(R.id.widgetStartWind)
        widgetwind3 = findViewById(R.id.widgetEndWind)
        widgetclouds1 = findViewById(R.id.widgetClouds1) // clouds
        widgetclouds2 = findViewById(R.id.widgetStartClouds)
        widgetclouds3 = findViewById(R.id.widgetEndClouds)
    }

    // Following the SETS of heads
    // date
    fun setWidgetDate1(title: String?) {
        widgetdate1.text = title
    }

    // start_tm
    fun setWidgetStartTm1(title: String?) {
        widgetstartTm1.text = title
    }

    // end_tm
    fun setWidgetEndTm1(title: String?) {
        widgetendTm1.text = title
    }

    // temperature
    fun setWidgetTemp1(title: String?) {
        widgettemp1.text = title
    }

    // wind
    fun setWidgetWind1(title: String?) {
        widgetwind1.text = title
    }

    // clouds
    fun setWidgetClouds1(title: String?) {
        widgetclouds1.text = title
    }

    // plausi for numeric input
    private val regEx = "^[0-9]*$"

    // following the GETS and SETS
    // get temperature with plausi
    var widgetTemp2: Int
        get() {
            val text = widgettemp2.text.toString()
            return if (isEmpty(text)) 0
            else if (!text.trim { it <= ' ' }
                    .matches(regEx.toRegex())) 100
            else {
                try {
                    text.replace("\\D".toRegex(), "").toInt()
                } catch (_: NumberFormatException) {
                    100
                }
            }
        }
        @SuppressLint("SetTextI18n")
        set(name) {
            if (name > 0 ) widgettemp2.setText(name.toString())
        }

    var widgetTemp3: Int
        get() {
            val text = widgettemp3.text.toString()
            return if (isEmpty(text)) 0
            else if (!text.trim { it <= ' ' }
                    .matches(regEx.toRegex())) 100
            else {
                try {
                    text.replace("\\D".toRegex(), "").toInt()
                } catch (_: NumberFormatException) {
                    100
                }
            }
        }
        @SuppressLint("SetTextI18n")
        set(name) {
            if (name > 0 ) widgettemp3.setText(name.toString())
        }

    // get wind with plausi
    var widgetWind2: Int
        get() {
            val text = widgetwind2.text.toString()
            return if (isEmpty(text)) 0
            else if (!text.trim { it <= ' ' }
                    .matches(regEx.toRegex())) 100
            else {
                try {
                    text.replace("\\D".toRegex(), "").toInt()
                } catch (_: NumberFormatException) {
                    100
                }
            }
        }

        @SuppressLint("SetTextI18n")
        set(name) {
            if (name > 0 ) widgetwind2.setText(name.toString())
        }

    var widgetWind3: Int
        get() {
            val text = widgetwind3.text.toString()
            return if (isEmpty(text)) 0
            else if (!text.trim { it <= ' ' }
                    .matches(regEx.toRegex())) 100
            else {
                try {
                    text.replace("\\D".toRegex(), "").toInt()
                } catch (_: NumberFormatException) {
                    100
                }
            }
        }
        @SuppressLint("SetTextI18n")
        set(name) {
            if (name > 0 ) widgetwind3.setText(name.toString())
        }

    // get clouds with plausi
    var widgetClouds2: Int
        get() {
            val text = widgetclouds2.text.toString()
            return if (isEmpty(text)) 0
            else if (!text.trim { it <= ' ' }
                    .matches(regEx.toRegex())) 200
            else {
                try {
                    text.replace("\\D".toRegex(), "").toInt()
                } catch (_: NumberFormatException) {
                    200
                }
            }
        }
        @SuppressLint("SetTextI18n")
        set(name) {
            if (name > 0 ) widgetclouds2.setText(name.toString())
        }

    var widgetClouds3: Int
        get() {
            val text = widgetclouds3.text.toString()
            return if (isEmpty(text)) 0
            else if (!text.trim { it <= ' ' }
                    .matches(regEx.toRegex())) 200
            else {
                try {
                    text.replace("\\D".toRegex(), "").toInt()
                } catch (_: NumberFormatException) {
                    200
                }
            }
        }
        @SuppressLint("SetTextI18n")
        set(name) {
            if (name > 0 ) widgetclouds3.setText(name.toString())
        }

    var widgetDate2: String?
        get() = widgetdate2.text.toString()
        set(name) {
            widgetdate2.text = name
        }
    var widgetStartTm2: String?
        get() = widgetstartTm2.text.toString()
        set(name) {
            widgetstartTm2.text = name
        }
    var widgetEndTm2: String?
        get() = widgetendTm2.text.toString()
        set(name) {
            widgetendTm2.text = name
        }

    companion object {
        /**
         * Checks if a CharSequence is empty ("") or null.
         *
         *
         * isEmpty(null)      = true
         * isEmpty("")        = true
         * isEmpty(" ")       = false
         * isEmpty("bob")     = false
         * isEmpty("  bob  ") = false
         *
         * @param cs the CharSequence to check, may be null
         * @return `true` if the CharSequence is empty or null
         */
        private fun isEmpty(cs: CharSequence?): Boolean {
            return cs.isNullOrEmpty()
        }
    }

}
